### PR TITLE
Allow compiling `#foo in obj` without compiling private fields

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

Currently we throw an error when compiling private brand checks without compiling private fields. This makes it impossible to add this plugin to `preset-env`, since it will throw an error in browsers with private fields support but without brand checks support.

This PR adds a new option to compile private brand checks without relying on the private fields transform: `preset-env` can then enable this option when necessary.

### CODE REVIEW

File: packages/babel-plugin-proposal-private-property-in-object/test/fixtures/to-native-fields/accessor/output.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+var _FooBrandCheck = new WeakSet();
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-private-property-in-object/test/fixtures/to-native-fields/accessor/output.js:1]: We can also place FooBrandCheck to a private static field, so we avoid the edge cases when `FooBrandCheck` might be injected to an incorrect scope, e.g. when a class is in the param initializer.

```js
// <-- FooBrandCheck will be injected to here
(x = class { #foo; test (other) { return #foo in other } }) => {}
```

File: packages/babel-plugin-proposal-private-property-in-object/test/fixtures/to-native-fields/accessor/output.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+class Foo {
+  constructor() {
+    _FooBrandCheck.add(this);
+  }
+
+  get #foo() {}
+
+  test(other) {
+    return _FooBrandCheck.has(other);
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-private-property-in-object/test/fixtures/to-native-fields/accessor/output.js:11]: Q: Is `_FooBrandCheck.has(other)` equivalent to `other instanceof Foo`?

Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-private-property-in-object/test/fixtures/to-native-fields/accessor/output.js:11]: No, it's different when other is a subclass of `Foo`

File: packages/babel-plugin-proposal-private-property-in-object/test/fixtures/to-native-fields/accessor/output.js (hunk 1 above)
Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-private-property-in-object/test/fixtures/to-native-fields/accessor/output.js:1]: Ohh I love this idea

File: packages/babel-plugin-proposal-private-property-in-object/test/fixtures/to-native-fields/accessor/output.js (hunk 2 above)
Comment (JLHwung) [reply packages/babel-plugin-proposal-private-property-in-object/test/fixtures/to-native-fields/accessor/output.js:11]: If `Bar`, a subclass of `Foo`, has `super()` in its constructor, does it imply `#foo in bar` is `true` where `bar` is an instance of `Bar`?

Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-private-property-in-object/test/fixtures/to-native-fields/accessor/output.js:11]: Oh I was confusing this test with the private fields one. `instanceof` is different because it checks the prototype:

```js
class Base {
  constructor() {
    return {};
  }
}

class Foo extends Base {
  #foo;

  static check(obj) {
    return #foo in obj;
  }
}

new Foo() instanceof Foo; // false
Foo.check(new Foo); // true
```

```js
class Foo {
  #foo;

  static check(obj) {
    return #foo in obj;
  }
}

let foo = new Foo();
foo.__proto__ = {};
foo instanceof Foo; // false
Foo.check(foo); // true
```

File: packages/babel-plugin-proposal-private-property-in-object/test/fixtures/to-native-fields/accessor/output.js (hunk 1 above)
Comment (JLHwung) [reply packages/babel-plugin-proposal-private-property-in-object/test/fixtures/to-native-fields/accessor/output.js:1]: I wish class access were not stage-2. The class binding can be overwritten. Guess we may have to stick to temporary variable.

**Review by jridgewell — COMMENTED** [review]
I think we're going to have ordering issues:

```js
class Foo {
  #x = 1;
  y = #x in this;
}

assert.equal(new Foo().y, true);
```

File: packages/babel-plugin-proposal-private-property-in-object/src/native-private-fields.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+          return;
+        }
+
+        if (isStatic) {
+          if (!outerClass.node.id) {
+            outerClass.set("id", path.scope.generateUidIdentifier("class"));
+          }
+          path.replaceWith(
+            template.expression.ast`
+              ${t.cloneNode(outerClass.node.id)} === ${path.node.right}
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-private-property-in-object/src/native-private-fields.js:76]: The class binding can be shadowed.

File: packages/babel-plugin-proposal-private-property-in-object/src/native-private-fields.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      ) {
+        firstFieldPath = el;
+        break;
+      }
+      if (!consturctorPath && el.isClassMethod({ kind: "constructor" })) {
+        consturctorPath = el;
+      }
+    }
+
+    if (firstFieldPath) {
```
Comment (jridgewell) [inline packages/babel-plugin-proposal-private-property-in-object/src/native-private-fields.js:31]: When doing `#foo in instance`, `instance` doesn't have the field until `#foo = init` completes. It's not correct to do this before the first field runs (the v8.dev [blog post](https://v8.dev/features/private-brand-checks) just reminded me of this, see the `halfConstructed` example).

Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-private-property-in-object/src/native-private-fields.js:31]: Nice I hate it

Comment (jridgewell) [reply packages/babel-plugin-proposal-private-property-in-object/src/native-private-fields.js:31]: Yes, accessors/methods are installed before the first initializer is run, so they cannot fail to install.

File: packages/babel-plugin-proposal-private-property-in-object/src/index.js [hunk 5]
```diff
@@ -6,15 +6,21 @@ import {
...
 } from "@babel/helper-create-class-features-plugin";
 
+import pluginPrivateIn from "./native-private-fields";
+
 export default declare((api, options) => {
   api.assertVersion(7);
 
+  const { loose, nativePrivateFields } = options;
+
+  if (nativePrivateFields) return pluginPrivateIn(api);
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-private-property-in-object/src/index.js:16]: can we default `nativePrivateFields` to `!isRequire("proposal-private-methods", api.targets())`?

**Review by jridgewell — APPROVED** [review]

File: packages/babel-plugin-proposal-private-property-in-object/src/native-private-fields.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+              outerClass.set("id", path.scope.generateUidIdentifier("class"));
+            }
+            path.replaceWith(
+              template.expression.ast`
+                ${t.cloneNode(outerClass.node.id)} === ${path.node.right}
+              `,
+            );
+          } else {
+            const id = getWeakSetId(
+              classWeakSets,
```
[the remark this answers is not in the document]
Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-private-property-in-object/src/native-private-fields.js:110]: This is used for instance methods, but it's a single weakset for all the instance methods in the class.

File: packages/babel-plugin-proposal-private-property-in-object/src/index.js (hunk 5 above)
Comment (hzoo) [reply packages/babel-plugin-proposal-private-property-in-object/src/index.js:16]: This isn't necessarily intended to be a public api right since we can just check targets? Or did we want to be able to specify that

Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-private-property-in-object/src/index.js:16]: Ideally every plugin should automatically compile as few as possible based on the targets, but users should still be able to override it via options.

**Comment by nicolo-ribaudo** [comment]
@JLHwung Rather than using `targets`, I realized that we can _always_ use the new implementation (https://github.com/babel/babel/pull/13172/commits/bf01056db4e7009ad292cbf78453f6d783bb84f3). All the tests in https://github.com/nicolo-ribaudo/babel/tree/private-in-to-native-private/packages/babel-plugin-proposal-private-property-in-object/test/fixtures/private ensure that we keep the old behavior when compiling private fields.

File: packages/babel-plugin-proposal-private-property-in-object/src/native-private-fields.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
+import syntaxPlugin from "@babel/plugin-syntax-private-property-in-object";
+import { injectInitialization as injectConstructorInit } from "@babel/helper-create-class-features-plugin";
+
+export default function pluginPrivateIn({ types: t, template }) {
+  const classWeakSets = new WeakMap();
+  const fieldsWeakSets = new WeakMap();
+
+  function unshadow(name, targetScope, scope) {
+    while (scope !== targetScope) {
+      if (scope.hasOwnBinding(name)) scope.rename(name);
```
[the remark this answers is not in the document]
Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-private-property-in-object/src/native-private-fields.js]: I think I'll leave this as-is: renaming a class is always observable (because it changes the `.name`), while renaming the other conflicts is often not observable (if we rename a variable). It can still be observable if the conflict is with a class/function, but at least it's not always.

**Comment by JLHwung** [comment]
@nicolo-ribaudo Does the new implementation increase the output size when all class features are compiled? If so I would prefer to use it only when private elements are natively supported.

**Comment by nicolo-ribaudo** [comment]
When all the class features are compiled we match the old output, because it's handled by the class features helper enabled by one of the other plugins. You can see that https://github.com/babel/babel/pull/13172/commits/bf01056db4e7009ad292cbf78453f6d783bb84f3 didn't change any test.

**Review by JLHwung — APPROVED** [review]
